UIKit/Location Controller/TWTLocationController.m

```diff
@@ -0,0 +1,189 @@
+        self.locationManager.desiredAccuracy = kCLLocationAccuracyHundredMeters;
+    }
+    return self;
+}
+
+
+#pragma mark - Fetch Location
+
+- (void)fetchCurrentLocationWithCompletion:(TWTLocationControllerCompletion)completion
+{
+    self.locationCompletion = completion;
+
+    self.requestBlocks = [[NSMutableArray alloc] init];
+    [self.requestBlocks addObject:completion];
+
+    [self updateLocationForAuthorizationStatus:[CLLocationManager authorizationStatus]];
+}
+
+
+#pragma mark - Validate Location
+
+- (BOOL)validateLocation:(CLLocation *)location accuracy:(CLLocationAccuracy)accuracy
+{
+    NSTimeInterval locationTimestampAgeInSeconds = [[NSDate date] timeIntervalSinceDate:location.timestamp];
+    return (locationTimestampAgeInSeconds < kLocationTimeoutInterval && location.horizontalAccuracy <= accuracy && location.horizontalAccuracy >= 0);
+}
+
+
+#pragma mark - Location Request Timeout
+
+- (void)locationUpdateTimedOut:(NSTimer *)timer
+{
+    if ([self validateLocation:self.location accuracy:kCLLocationAccuracyThreeKilometers]) {
+        [self processCompletionBlocksWithLocation:self.location error:nil];
+        [self stopUpdatingLocation];
+    }
+    else {
+        NSLog(@"location timed out.");
+        [self failWithErrorCode:TWTLocationControllerErrorUpdateTimedOut underlyingError:nil];
+    }
+}
+
+
+#pragma mark - CLLocationManagerDelegate
+
+- (void)locationManager:(CLLocationManager *)manager didChangeAuthorizationStatus:(CLAuthorizationStatus)status
+{
+    [self updateLocationForAuthorizationStatus:status];
+}
+
+
+- (void)locationManager:(CLLocationManager *)manager didUpdateLocations:(NSArray *)locations
+{
+    self.location = [locations lastObject];
+
+    if ([self validateLocation:self.location accuracy:manager.desiredAccuracy]) {
+        [self processCompletionBlocksWithLocation:self.location error:nil];
+        [self stopUpdatingLocation];
+    }
+}
+
+
+- (void)locationManager:(CLLocationManager *)manager didFailWithError:(NSError *)error
+{
+    // TODO (foster): What to pass here as an error code.
+    [self failWithErrorCode:error.code underlyingError:error];
+}
+
+
+#pragma mark - Helpers
+
+- (void)updateLocationForAuthorizationStatus:(CLAuthorizationStatus)status
+{
+    switch ([CLLocationManager authorizationStatus]) {
+
+        case kCLAuthorizationStatusNotDetermined: {
+            [self failWithErrorCode:TWTLocationControllerAuthorizationStatusNotDetermined underlyingError:nil];
+            [self.locationManager requestWhenInUseAuthorization];
+            [self.locationManager requestAlwaysAuthorization];
+            break;
+        }
+
+        case kCLAuthorizationStatusRestricted: {
+            [self failWithErrorCode:TWTLocationControllerAuthorizationStatusRestricted underlyingError:nil];
+            break;
+        }
+
+        case kCLAuthorizationStatusDenied: {
+            [self failWithErrorCode:TWTLocationControllerErrorPermissionDenied underlyingError:nil];
```

Comment:
Seems like these cases should basically be wrapped up in a `TWTLocationControllerErrorCodeNotAuthorized` and you should pass in the actual status as a user info key (which should be documented and placed in the header).